enhancements/cloud-integration/openstack/failure-domains.md

```diff
@@ -0,0 +1,715 @@
+
+Note that in deployments where all the servers are spawned in Failure domains, and where all these Failure domains have Ports specified, the `MachinesSubnet` may remain empty. To prevent the Installer from creating a `MachinesSubnet` subnet, set `MachinesSubnet` in install-config to an existing subnet.
+
+### OpenStackProviderSpec
+
+OpenStackProviderSpec includes a copy of the failure domain. This means that a higher level orchestrator, e.g. the installer when creating multiple MachineSets, or CPMS when creating Machines in different failure domains, only needs to do a simple copy. The Machine controller will validate and merge the FailureDomain with the rest of the MachineSpec at server creation time.
+
+```go
+    // The OpenStackFailureDomain used to create the server.
+    FailureDomains OpenStackFailureDomain `json:"failureDomains,omitempty"`
+```
+
+### Control Plane MachineSet
+
+The full details of integration with Control Plane MachineSet are out of scope of this enhancement. However, we must ensure this enhancement is compatible with future integration with CPMS.
+
+Our primary concern is to ensure we can implement [InjectFailureDomain and ExtractFailureDomain from ProviderConfig](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/7f66c3455bda9b23ad063eafd4c797541e358d6d/pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig.go#L51-L59):
+
+```go
+  type ProviderConfig interface {
+    // InjectFailureDomain is used to inject a failure domain into the ProviderConfig.
+    // The returned ProviderConfig will be a copy of the current ProviderConfig with
+    // the new failure domain injected.
+    InjectFailureDomain(failuredomain.FailureDomain) (ProviderConfig, error)
+
+    // ExtractFailureDomain is used to extract a failure domain from the ProviderConfig.
+    ExtractFailureDomain() failuredomain.FailureDomain
+```
+
+We can always do this unambiguously because we store the entire failure domain in the ProviderSpec.
+
+### Workflow Description
+
+#### Create a new cluster with control plane and worker failure domains
+
+An OpenStack cloud has 3 failure domains: az1, az2, and az3. Each of these has corresponding Nova(compute) and Cinder(storage) availability zones of the same name. Additionally each failure domain contains a pre-created provider network with a subnet.
+
+The user wants to create a control plane with 3 nodes with each node in a different failure domain, and 6 workers distributed evenly across the same 3 failure domains.
+
+The user creates an install config containing the following:
+
+```yaml
+platform:
+  openstack:
+    machinesSubnet: 31ad4088-7e52-48da-9df5-35e70da438f6
+controlPlane:
+  name: control-plane
+  platform:
+    openstack:
+      type: m1.large
+      rootVolume:
+        size: 50
+        type: gold
+      failureDomains:
```

Comment:
Should all resources on the same failureDomain be in the same az? If not, perhaps we should mention that

Reply:
In our proposed definition, a failure domain is a bundle of:
* zero or one Compute zones
* zero or one Storage zones
* zero or one subnets

I am adding this to the doc for clarity. Is this answering your question?

Reply:
Where did that get answered, because while I see your definition, I don't think it answers Maysa's question. If the failure domain is defined as no compute zone, does that mean that the `AvailabilityZone` must be empty? (Which IIUC means OpenStack picks the AZ when you create the VM?)